core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -1871,6 +1886,225 @@ void skip0() {
         private static final int KERN_PROCARGS2 = 49;
     }
 
+    /**
+     * Implementation for FreeBSD based on sysctl(3).
+     */
+    private static class FreeBSD extends Unix {
+
+        // Taken from sys/errno.h
+        private static final int ENOMEM = 12;
+
+        // Taken from sys/sysctl.h
+        private static final int CTL_KERN = 1;
+        private static final int KERN_ARGMAX = 8;
+        private static final int KERN_PROC = 14;
+        private static final int KERN_PROC_ALL = 0;
+        private static final int KERN_PROC_ARGS = 7;
+        private static final int KERN_PROC_ENV = 35;
+
+        // Local constants
+        private final long sizeOf_kinfo_proc;
+        private static final long sizeOf_kinfo_proc_32 = 768;
+        private static final long sizeOf_kinfo_proc_64 = 1088;
+        private final int kinfo_proc_pid_offset;
+        private static final int kinfo_proc_pid_offset_32 = 40;
+        private static final int kinfo_proc_pid_offset_64 = 72;
+        private final int kinfo_proc_ppid_offset;
+        private static final int kinfo_proc_ppid_offset_32 = 44;
+        private static final int kinfo_proc_ppid_offset_64 = 76;
+        private static final int sizeOfInt = Native.getNativeSize(int.class);
+
+        FreeBSD(boolean vetoersExist) {
+            super(vetoersExist);
+
+            String arch = System.getProperty("sun.arch.data.model");
+            if ("64".equals(arch)) {
+                sizeOf_kinfo_proc = sizeOf_kinfo_proc_64;
+                kinfo_proc_pid_offset = kinfo_proc_pid_offset_64;
+                kinfo_proc_ppid_offset = kinfo_proc_ppid_offset_64;
+            } else {
+                sizeOf_kinfo_proc = sizeOf_kinfo_proc_32;
+                kinfo_proc_pid_offset = kinfo_proc_pid_offset_32;
+                kinfo_proc_ppid_offset = kinfo_proc_ppid_offset_32;
+            }
+            try {
+                NativeLongByReference size = new NativeLongByReference(new NativeLong(0));
+                Memory m;
+                int nRetry = 0;
+                while (true) {
+                    // Find out how much memory we need for kern.proc.all.
+                    if (LIBC.sysctl(
+                                    new int[] {CTL_KERN, KERN_PROC, KERN_PROC_ALL},
+                                    3,
+                                    NULL,
+                                    size,
+                                    NULL,
+                                    new NativeLong(0))
+                            != 0) {
+                        throw new IOException(
+                                "Failed to get memory requirement: "
+                                        + LIBC.strerror(Native.getLastError()));
+                    }
+
+                    // Add some padding to account for new processes.
+                    long len = size.getValue().longValue();
+                    len += len / 10L;
```

Comment:
Any reasoning behind this exact number?

Reply:
I saw some other code on the Internet bit-shifting by 3 or padding by 10%. Intuitively it makes sense to add some padding to account for new processes that may have launched during the time between the first call and the second. The exact number doesn't matter much because we retry anyway. And on a system with a large amount of short-lived processes, no amount of retrying will save you. We don't even pad at all on macOS and nobody seems to have complained (though maybe that could be an RFE for the macOS implementation).